navigation/localization/auv_localization/auv_localization/auv_localization.py

```diff
@@ -0,0 +1,220 @@
+
+        # Kalman Filter noise parameters
+        self.declare_parameter("Q_noise", 10 ** -3)
+        self.declare_parameter("R_noise", 10 ** -3)
+        self.declare_parameter("R_noise_depth", 10 ** -6)
+        self.declare_parameter("P_noise", 10 ** -2)
+
+    def drone_est_position_cb(self, msg: Odometry):
+        self.drone_estimated_position = np.array([msg.pose.pose.position.x,
+                                                  msg.pose.pose.position.y,
+                                                  msg.pose.pose.position.z])
+    
+    def drone_ground_truth_cb(self, msg: Odometry):
+        # Store the latest ground truth position
+        self.drone_ground_truth_position = np.array([msg.pose.pose.position.x,
+                                                     msg.pose.pose.position.y,
+                                                     msg.pose.pose.position.z])
+
+    def depth_cb(self, msg: Float32) -> None:
+        self.depth = msg.data
+
+    def min_gradient_cb(self, msg: Float32MultiArray) -> None:
+        u, v = msg.data[0], msg.data[1]
+        self.image_point = np.array([u, v, 1])
+        
+        if self.image_point is not None: 
+            # and self.X_drone is not None:
+            self.estimate_relative_position()
+
+    def estimate_relative_position(self):
+        observation = self.image_point.copy()
+        observation[2] = self.depth
+
+        if not self.first_measurement:
+            self.X_auv_relative = self.get_first_measured_state(observation)
+            self.kf_auv.initialize_state(self.X_auv_relative, self.Q_auv_relative, self.R_auv_relative, self.P_auv_relative)
+            self.first_measurement = True
+            # self.get_logger().info(f"initializaing: {self.X_auv_relative }") debugging....
+        else:
+            self.X_auv_relative, self.P_auv_relative = self.kf_auv.estimate(observation, feedback_image=True)
+            self.get_logger().info(f"estimating relative position: {self.X_auv_relative }")
+        if self.X_auv_relative is not None:
+            auv_position = self.get_world_position(self.X_auv_relative[:3])
+            self.publish_auv_position(auv_position)
+            self.publish_tf(auv_position)
+
+    def get_first_measured_state(self, observation):
+        direction_vector = np.dot(self.K_inv, observation)
+        direction_world = np.dot(np.eye(3), direction_vector)  # Using an identity matrix for R_wa
+        scale = self.depth / direction_world[2]
+        vector_auv_drone = direction_world * scale
+        return np.array([vector_auv_drone, np.zeros(3)]).flatten().T
+
+    def publish_tf(self, position, orientation=None):
+        # Create a TransformStamped message
+        t = TransformStamped()
+
+        # Set the frame information
+        t.header.stamp = self.get_clock().now().to_msg()
+        t.header.frame_id = self.map_frame  # Parent frame
+        t.child_frame_id = 'auv_estimated_state'  # Child frame for the estimated state
```

Comment:
is the auv estimated state wrt to map frame or drone's odom frame? the odom publisher below suggests odom, this suggests map, they both use the same _numbers_. seems a little wonky.

Reply:
it do be in the map frame because I first get relative position in map_frame and add it it to the dead_reckoning/odom_gt subscribed coordinate, which is also in map frame

Reply:
then just make sure the parent frames are consistent :)